index.d.ts

```diff
@@ -0,0 +1,208 @@
+/// <reference types="connect" />
+
+import { IncomingMessage, ServerResponse } from 'http';
+import { Handler } from 'aws-lambda';
+import { ErrorHandleFunction } from 'connect';
+
+export default agent;
+
+declare const agent: Agent;
+
+declare class Agent implements Taggable, StartSpanFn {
+  // Configuration
+  start (options?: AgentConfigOptions): Agent;
+  isStarted (): boolean;
+
+  // Data collection hooks
+  middleware: { connect (): ErrorHandleFunction };
+  lambda (handler: Handler): Handler;
+  lambda (type: string, handler: Handler): Handler;
+  handleUncaughtExceptions (
+    fn: (err: Error) => void
+  ): void;
+
+  // Errors
+  captureError (
+    err: Error | string | ParameterizedMessageObject,
+    callback?: CaptureErrorCallback
+  ): void;
+  captureError (
+    err: Error | string | ParameterizedMessageObject,
+    options?: CaptureErrorOptions,
+    callback?: CaptureErrorCallback
+  ): void;
+
+  // Transactions
+  startTransaction (
+    name?: string,
+    type?: string,
+    options?: TransactionOptions
+  ): Transaction | null;
+  setTransactionName (name: string): void;
+  endTransaction (result?: string, endTime?: number): void; // TODO: Should we allow number as well for result?
+  currentTransaction: Transaction | null;
+
+  // Spans
+  startSpan (name?: string, type?: string, options?: SpanOptions): Span | null;
+  currentSpan: Span | null;
+
+  // Context
+  setTag (name: string, value: TagValue): boolean;
+  addTags (tags: Tags): boolean;
+  setUserContext (user: UserObject): void;
+  setCustomContext (custom: object): void;
+
+  // Transport
+  addFilter (fn: FilterFn): void;
+  addErrorFilter (fn: FilterFn): void;
+  addSpanFilter (fn: FilterFn): void;
+  addTransactionFilter (fn: FilterFn): void;
+  flush (callback: Function): void;
+  destroy (): void;
+
+  // Utils
+  logger: Logger;
+}
+
+declare class GenericSpan implements Taggable {
+  // TODO: The following should not be documented right? constructor(), timestamp, ended, id, traceId, parentId, sampled, duration()
+  type: string | null; // TODO: Should we allow null?
+
+  setTag (name: string, value: TagValue): boolean;
+  addTags (tags: Tags): boolean;
+}
+
+declare class Transaction extends GenericSpan implements StartSpanFn {
+  // TODO: The following should not be documented right? constructor(), setUserContext(), setCustomContext(), toJSON(), setDefaultName(), setDefaultNameFromRequest()
+  name: string | null; // TODO: Should we allow null?
+  result: string; // TODO: Should we also document number?
+
+  startSpan (name?: string, type?: string, options?: SpanOptions): Span | null;
+  ensureParentId (): string;
+  end (result?: string, endTime?: number): void; // TODO: Should we allow number as well for result?
+}
+
+declare class Span extends GenericSpan {
+  // TODO: The following should not be documented right? constructor(), customStackTrace(), setDbContext()
+  transaction: Transaction;
+  name: string | null; // TODO: Should we allow null?
```

Comment:
Same as GenericSpan type, I think it's valid for a Span to have a null name. We apply a fallback internally, but that's not exposed to the outer surface of the object to work for any runtime checks one might try to do to validate this.